src/main/java/org/wildfly/security/http/HttpAuthenticator.java

```diff
@@ -98,12 +110,20 @@ public static Builder builder() {
         private volatile HttpServerMechanismsResponder successResponder;
 
         private boolean authenticate() throws HttpAuthenticationException {
+            return evaluate((ExceptionConsumer<HttpServerAuthenticationMechanism, HttpAuthenticationException>) mechanism -> mechanism.evaluateRequest(this));
+        }
+
+        private boolean logout() throws HttpAuthenticationException {
+            return evaluate((ExceptionConsumer<HttpServerAuthenticationMechanism, HttpAuthenticationException>) mechanism -> mechanism.evaluateLogoutRequest(this));
+        }
+
+        private boolean evaluate(ExceptionConsumer<HttpServerAuthenticationMechanism, HttpAuthenticationException> mechanism) throws HttpAuthenticationException {
             List<HttpServerAuthenticationMechanism> authenticationMechanisms = mechanismSupplier.get();
             responders = new ArrayList<>(authenticationMechanisms.size());
             try {
                 for (HttpServerAuthenticationMechanism nextMechanism : authenticationMechanisms) {
                     currentMechanism = nextMechanism;
```

Comment:
The logout is programmatic, thus a result of a notification from the application/container. We don't really care about the pushing things out to the response because logout will usually be performed latter after.

The point here is that we allow mechanisms to clean up an existing authentication session, where mechanisms can decide whether the logout request really comes from an authenticated user (by checking the state associated with that user in a mech specific way) or not and acting accordingly.

This also allows applications to do whatever they want after a programmatic logout, like redirect user to some page or just force a refresh in order to redirect to login page. It is up to the app, not the mech to decide about the flow. What I think is a good thing.